core/modules/ip_defrag.cc

```diff
@@ -0,0 +1,128 @@
+enum {DEFAULT_GATE = 0, FORWARD_GATE};
+/*----------------------------------------------------------------------------------*/
+/**
+ * Returns NULL if packet is fragmented and needs more for reassembly.
+ * Returns Packet ptr if the packet is unfragmented, or is freshly reassembled.
+ */
+bess::Packet *
+IPDefrag::IPReassemble(Context *ctx, bess::Packet *p)
+{
+	Ethernet *eth = p->head_data<Ethernet *>();
+	if (eth->ether_type != (be16_t)(Ethernet::kIpv4))	return p;
+	Ipv4 *iph = (Ipv4 *)((unsigned char *)eth + sizeof(Ethernet));
+
+	if (rte_ipv4_frag_pkt_is_fragmented((struct rte_ipv4_hdr *)iph)) {
+		struct rte_mbuf *mo, *m;
+		struct rte_ipv4_hdr *ip;
+
+		/* prepare mbuf: setup l2_len/l3_len */
+		m = reinterpret_cast<struct rte_mbuf *>(p);
+		ip = reinterpret_cast<struct rte_ipv4_hdr *>(iph);
+		m->l2_len = sizeof(*eth);
+		m->l3_len = sizeof(*iph);
+
+		/* process this fragment */
+		mo = rte_ipv4_frag_reassemble_packet(ift, &ifdr, m, cur_tsc, ip);
+		if (mo == NULL) {
+			/* no packet to process just yet */
+			p = NULL;
+			return p;
+		}
+		/* we have our packet reassembled */
+		if (mo != m) {
+			/* move mbuf data in the first segment */
+			if (rte_pktmbuf_linearize(mo) == 0) {
+				p = reinterpret_cast<bess::Packet *>(mo);
+				eth = p->head_data<Ethernet *>();
+				iph = (Ipv4 *)((unsigned char *)eth + sizeof(Ethernet));
+			} else {
+				DLOG(INFO) << "Failed to linearize rte_mbuf. "
+					   << "Is there enough tail room?"
+					   << std::endl;
+				EmitPacket(ctx, p, DEFAULT_GATE);
+				return NULL;
+			}
+		}
+	}
+	/* Reset checksum. This will be computed by a later module in line (if needed) */
+	iph->checksum = 0;
+
+	return p;
+}
+/*----------------------------------------------------------------------------------*/
+void
+IPDefrag::ProcessBatch(Context *ctx, bess::PacketBatch *batch)
+{
+	/* retire outdated frags (if needed) */
+	if (ifdr.cnt != 0)
+		rte_ip_frag_free_death_row(&ifdr, PREFETCH_OFFSET);
+	cur_tsc = rte_rdtsc();
+
+	int cnt = batch->cnt();
+	for (int i = 0; i < cnt; i++) {
+		bess::Packet *p = batch->pkts()[i];
+		p = IPReassemble(ctx, p);
+		if (p) EmitPacket(ctx, p, FORWARD_GATE);
+	}
+}
+/*----------------------------------------------------------------------------------*/
+void
+IPDefrag::DeInit()
+{
+	if (ift != NULL) {
+		/* free allocated IP frags */
+		rte_ip_frag_table_destroy(ift);
+		ift = NULL;
+	}
+}
+/*----------------------------------------------------------------------------------*/
+CommandResponse
+IPDefrag::Init(const bess::pb::IPDefragArg &arg) {
+
+	num_flows = arg.num_flows();
+
+	if (num_flows <= 0)
+		return CommandFailure(EINVAL,
+				      "Invalid num_flows!");
+
+	frag_cycles = (rte_get_tsc_hz() + MS_PER_S - 1)
+		/ MS_PER_S * num_flows;
+
+	cur_tsc = 0;
+
+	ift = rte_ip_frag_table_create(num_flows,
+				       IP_FRAG_TBL_BUCKET_ENTRIES,
+				       num_flows * IP_FRAG_TBL_BUCKET_ENTRIES,
+				       frag_cycles,
+				       SOCKET_ID_ANY);
```

Comment:
Can we allow this to be set from the args? The default value being SOCKET_ID_ANY. That would help control program to be intelligent to use worker's socket id and should that fail, try another socket id?

Reply:
Done.